### PR TITLE
Enum Refactor: DataZoneEquipment

### PR DESCRIPTION
**Pull request by dareumnam**
Pull request overview
---------------------
Remove `constexpr int`s in DataZoneEquipment object

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/BaseboardElectric.cc [hunk 1]
```diff
@@ -338,7 +337,8 @@ namespace BaseboardElectric {
 
                 for (CtrlZone = 1; CtrlZone <= state.dataGlobal->NumOfZones; ++CtrlZone) {
                     for (ZoneEquipTypeNum = 1; ZoneEquipTypeNum <= state.dataZoneEquip->ZoneEquipList(CtrlZone).NumOfEquipTypes; ++ZoneEquipTypeNum) {
-                        if (state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipType_Num(ZoneEquipTypeNum) == BBElectricConvective_Num &&
+                        if (state.dataZoneEquip->ZoneEquipList(CtrlZone).EquipType_Num(ZoneEquipTypeNum) ==
```
Comment (amirroth) [inline src/EnergyPlus/BaseboardElectric.cc]: I know this is pedantic, but we should call things `Type` rather than `TypeNum`. Not so critical here, more important for member names.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 2]
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
...
+        Intermediate,
+        Outlet
+    };
 
-    constexpr int ZoneSplitter_Type(1);
-    constexpr int ZoneSupplyPlenum_Type(2);
-    constexpr int ZoneMixer_Type(3);
-    constexpr int ZoneReturnPlenum_Type(4);
+    enum class CompType
+    {
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh:90]: Not sure `CompType` is a good name for this `enum`. Also, need to add `Num` to the ends of these.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 3]
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
...
-    constexpr int CoolingPanel_Num(29);
-    constexpr int ZoneUnitarySys_Num(30);
-
-    // Per Person Ventilation Rate Mode
-    constexpr int PerPersonDCVByCurrentLevel(1);
-    constexpr int PerPersonByDesignLevel(2);
+    enum ZoneEquip
+    {
+        Unassigned = -1,
+        FanCoil4Pipe = 1,
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh:107]: Why is this not 0? Is it because there are `Array1D` that use these as indices? That should be a sign to switch those to `std::array`. Can do this in a subsequent PR I guess.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 4]
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
...
+        HeatXchngr,
+        HPWaterHeater,
+        BBWater,
+        ZoneDXDehumidifier,
+        BBSteam,
+        BBElectric,
+        RefrigerationAirChillerSet,
+        UserDefinedZoneHVACForcedAir,
+        CoolingPanel,
+        ZoneUnitarySys
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh]: Add `Num`.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 5]
```diff
@@ -76,57 +76,73 @@ namespace DataZoneEquipment {
...
     constexpr int NumValidSysAvailZoneComponents(14);
     extern Array1D_string const cValidSysAvailManagerCompTypes;
 
+    // Per Person Ventilation Rate Mode
+    enum class PerPersonVentRateMode
+    {
+        Unassgined = -1,
+        DCVByCurrentLevel,
+        ByDesignLevel
+    };
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh:151]: Add `Num`.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 6]
```diff
@@ -358,7 +374,7 @@ namespace DataZoneEquipment {
         int NumAvailHeatEquip;                      // Number of pieces of equipment available for heating
         int NumAvailCoolEquip;                      // Number of pieces of equipment available for cooling
         Array1D_string EquipType;
-        Array1D_int EquipType_Num;
+        Array1D<DataZoneEquipment::ZoneEquip> EquipType_Num;
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh]: No `_Num`. I understand that there is already an array called `EquipType` which is an array of strings, but that array should not be necessary. You should convert things from string to `enum` as we parse the IDF and only use `enum` internally. If you want, you can put a pin in these comments and do them in a separate PR.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 7]
```diff
@@ -407,7 +423,7 @@ namespace DataZoneEquipment {
         int NumOfComponents;
         int InletNodeNum;
         Array1D_string ComponentType;
-        Array1D_int ComponentType_Num;
+        Array1D<DataZoneEquipment::CompType> ComponentType_Num;
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh]: Same comment as above.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 8]
```diff
@@ -431,7 +447,7 @@ namespace DataZoneEquipment {
         int NumOfComponents;
         int OutletNodeNum;
         Array1D_string ComponentType;
-        Array1D_int ComponentType_Num;
+        Array1D<DataZoneEquipment::CompType> ComponentType_Num;
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.hh]: Same comment.

File: src/EnergyPlus/FanCoilUnits.cc [hunk 9]
```diff
@@ -712,7 +712,7 @@ namespace FanCoilUnits {
                        state.dataFanCoilUnits->ATMixerOutNode,
                        FanCoil(FanCoilNum).AirOutNode);
             FanCoil(FanCoilNum).ControlZoneNum =
-                DataZoneEquipment::GetZoneEquipControlledZoneNum(state, DataZoneEquipment::FanCoil4Pipe_Num, FanCoil(FanCoilNum).Name);
+                DataZoneEquipment::GetZoneEquipControlledZoneNum(state, DataZoneEquipment::ZoneEquip::FanCoil4Pipe, FanCoil(FanCoilNum).Name);
```
Comment (amirroth) [inline src/EnergyPlus/FanCoilUnits.cc:712]: Probably not something you want to handle in this PR, but ... why are we doing this lookup by FanCoil name if we already know FanCoil index?

File: src/EnergyPlus/DataZoneEquipment.cc [hunk 10]
```diff
@@ -516,107 +516,107 @@ void GetZoneEquipmentData(EnergyPlusData &state)
                     auto const SELECT_CASE_var(UtilityRoutines::MakeUPPERCase(thisZoneEquipList.EquipType(ZoneEquipTypeNum)));
 
                     if (SELECT_CASE_var == "ZONEHVAC:AIRDISTRIBUTIONUNIT") {
```
Comment (amirroth) [inline src/EnergyPlus/DataZoneEquipment.cc:518]: This code should be converted to a `getEnumerationValue`.

File: src/EnergyPlus/ReturnAirPathManager.cc [hunk 11]
```diff
@@ -191,9 +188,9 @@ namespace ReturnAirPathManager {
                             ErrorsFound = true;
                         }
                         if (UtilityRoutines::SameString(state.dataIPShortCut->cAlphaArgs(Counter), "AirLoopHVAC:ZoneMixer"))
-                            state.dataZoneEquip->ReturnAirPath(PathNum).ComponentType_Num(CompNum) = ZoneMixer_Type;
+                            state.dataZoneEquip->ReturnAirPath(PathNum).ComponentType_Num(CompNum) = DataZoneEquipment::CompType::ZoneMixer;
```
Comment (amirroth) [inline src/EnergyPlus/ReturnAirPathManager.cc]: `getEnumerationValue`.

File: src/EnergyPlus/SystemReports.cc [hunk 12]
```diff
@@ -4539,7 +4539,7 @@ void ReportMaxVentilationLoads(EnergyPlusData &state)
                                                .EquipType_Num(thisZoneEquipNum));
                 // case statement to cover all possible zone forced air units that could have outside air
 
-                if (SELECT_CASE_var == WindowAC_Num) { // Window Air Conditioner
+                if (SELECT_CASE_var == DataZoneEquipment::ZoneEquip::WindowAC) { // Window Air Conditioner
```
Comment (amirroth) [inline src/EnergyPlus/SystemReports.cc]: This should be a `switch`/`case`.

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 13]
```diff
@@ -3065,7 +3065,7 @@ void SimZoneEquipment(EnergyPlusData &state, bool const FirstHVACIteration, bool
             {
                 auto const SELECT_CASE_var(state.dataZoneEquip->SupplyAirPath(SupplyAirPathNum).ComponentType_Num(CompNum));
 
-                if (SELECT_CASE_var == ZoneSplitter_Type) { // 'AirLoopHVAC:ZoneSplitter'
+                if (SELECT_CASE_var == DataZoneEquipment::CompType::ZoneSplitter) { // 'AirLoopHVAC:ZoneSplitter'
```
Comment (amirroth) [inline src/EnergyPlus/ZoneEquipmentManager.cc]: `switch`/`case`.

File: src/EnergyPlus/FanCoilUnits.cc (hunk 9 above)
Comment (rraustad) [reply src/EnergyPlus/FanCoilUnits.cc:712]: ZoneEquipConfig stores the component type and name so to look up the correct component needed the type and name (back then). Now that code is progressing a new method might be used.

**Comment by Myoldmopar** [comment]
Looks like comments have been addressed here, but speak up if there is anything still waiting.  I pulled develop into this locally and re-ran regressions.  Everything passes beautifully.  If no one has an issue, I'll merge this this afternoon.

File: src/EnergyPlus/AirflowNetworkBalanceManager.cc [hunk 14]
```diff
@@ -12325,9 +12325,6 @@ namespace AirflowNetworkBalanceManager {
         // PURPOSE OF THIS SUBROUTINE:
         // This subroutine validate zone exhaust fan and associated surface
 
-        // Using/Aliasing
-        using DataZoneEquipment::ZoneExhaustFan_Num;
```
Comment (Myoldmopar) [inline src/EnergyPlus/AirflowNetworkBalanceManager.cc:12329]: A simple but nice change.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 15]
```diff
@@ -5823,9 +5823,9 @@ void DynamicIntConvSurfaceClassification(EnergyPlusData &state, int const SurfNu
                  EquipNum <= state.dataZoneEquip->ZoneEquipList(state.dataZoneEquip->ZoneEquipConfig(ZoneNum).EquipListIndex).NumOfEquipTypes;
                  ++EquipNum) {
 
-                switch (state.dataZoneEquip->ZoneEquipList(state.dataZoneEquip->ZoneEquipConfig(ZoneNum).EquipListIndex).EquipType_Num(EquipNum)) {
-                case AirDistUnit_Num:
-                case PurchasedAir_Num:
+                switch (state.dataZoneEquip->ZoneEquipList(state.dataZoneEquip->ZoneEquipConfig(ZoneNum).EquipListIndex).EquipTypeEnum(EquipNum)) {
+                case DataZoneEquipment::ZoneEquip::AirDistUnit:
+                case DataZoneEquipment::ZoneEquip::PurchasedAir:
```
Comment (Myoldmopar) [inline src/EnergyPlus/ConvectionCoefficients.cc:5828]: This reads so much nicer.

File: src/EnergyPlus/DataZoneEquipment.hh [hunk 16]
```diff
@@ -76,63 +76,87 @@ namespace DataZoneEquipment {
...
-    constexpr int Outlet(4);
+    enum class AirNodeType
+    {
+        Unassigned = -1,
+        PathInlet,
+        CompInlet,
+        Intermediate,
+        Outlet,
+        NUM
+    };
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataZoneEquipment.hh:87]: Awesome.

File: src/EnergyPlus/DataZoneEquipment.hh (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/DataZoneEquipment.hh:107]: @dareumnam any comment on this?

**Review by Myoldmopar — APPROVED** [review]
Everything looks OK to me.  There appears to be one unanswered question which I noted in my comments, but even that is something for later.  I think this is good to go.

File: src/EnergyPlus/DataZoneEquipment.hh (hunk 3 above)
Comment (rraustad) [reply src/EnergyPlus/DataZoneEquipment.hh:107]: image

Comment (amirroth) [reply src/EnergyPlus/DataZoneEquipment.hh:107]: Per my comment above, arrays indexed by `enum` should be turned into `std::array`. Per same comment, no need to hold up this PR for that single change.

**Comment by dareumnam** [comment]
@Myoldmopar Thanks for your review. I keep addressing comments on my local branch, but it is okay if this gets merged in first. I can open a separate PR.

File: src/EnergyPlus/DataZoneEquipment.hh (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/DataZoneEquipment.hh:107]: :+1:

**Comment by Myoldmopar** [comment]
Alright, pulled develop in just for fun after the version change and it's still passing fine.  @dareumnam just make sure to address the last little pieces here on your next PR.  Thanks!  Merging!
